File: new/cpp/704_binary_search_built_in.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // idea: use built in functions (iterators)

        // this will find out the first lowest address where we can insert the target
        auto iterator = lower_bound(nums.begin(), nums.end(), target);
        return ((iterator != nums.end() && *iterator == target) ? iterator - nums.begin() : - 1);
    }
    // time: O(logn)
    // space: O(1)
};

File: new/python/704_binary_search_built_in.py
import bisect


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # idea: using built in functions
        index = bisect.bisect_left(nums, target)
        return index if (index < len(nums) and nums[index] == target) else -1

    # time: O(logn) - internally bisect.bisect_left uses binary search (lower bound)
    # space: O(1)
